File: leetcode/222.完全二叉树的节点个数/222.完全二叉树的节点个数.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func countNodes(root *TreeNode) int {
	// 记录左右子树的高度
	left := root
	right := root
	leftHigh := 0
	rightHigh := 0
	for left != nil {
		left = left.Left
		leftHigh += 1
	}
	for right != nil {
		right = right.Right
		rightHigh += 1

	}
	// 左右子树高度相等是是满二叉树
	// 满二叉树情况下 节点总数 = 2^high - 1
	if leftHigh == rightHigh {
		return IntPow(2, leftHigh) - 1
	}
	// 普通二叉树递归计算，1为当前节点数量
	return 1 + countNodes(root.Left) + countNodes(root.Right)
}

func IntPow(n, m int) int {
	if m == 0 {
		return 1
	}
	result := n
	for i := 2; i <= m; i++ {
		result *= n
	}
	return result
}

File: leetcode/222.完全二叉树的节点个数/222.完全二叉树的节点个数.py
import math


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def countNodes(self, root: TreeNode) -> int:
        # 求左右子树高度
        left = right = root
        left_high = right_high = 0
        while left:
            left = left.left
            left_high += 1
        while right:
            right = right.right
            right_high += 1
        # 如果左右子树相同是满二叉树
        # 满二叉树节点总数 = 2^high -1
        if left_high == right_high:
            return int(math.pow(2, left_high)) - 1
        # 普通二叉树进入递归相加，1是为当前节点
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)
